File: Планшет/app/src/main/java/ua/church/slovo/remote/Library.java
package ua.church.slovo.remote;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteStatement;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/// Своя бібліотека планшета: переклади й пісенники, з якими проповідник
/// складає план удома, без зв'язку з програмою.
///
/// Звідки б модуль не прийшов — зі «Слова» по Wi-Fi, з GitHub чи з файла на
/// планшеті, — він лягає в одну базу однаковими рядками: книга, розділ, вірш,
/// текст. Екран плану тому не знає, якого формату був модуль.
///
/// Разом із модулем з GitHub чи з файла зберігається й сам файл (`original`):
/// якщо в програмі на служінні такого перекладу не виявиться, планшет відвезе
/// їй саме його, і програма поставить модуль собі.
final class Library extends SQLiteOpenHelper {

    static final String BIBLE = "bible";
    static final String SONGS = "songs";

    static final class Module {
        final String id;
        final String kind;
        final String name;
        final String shortName;
        /// «slovo», «github» чи «file».
        final String source;
        /// Шлях до збереженого файла модуля; порожньо — модуль прийшов зі «Слова».
        final String original;

        Module(String id, String kind, String name, String shortName, String source, String original) {
            this.id = id;
            this.kind = kind;
            this.name = name;
            this.shortName = shortName;
            this.source = source;
            this.original = original;
        }
    }

    static final class Book {
        final int index;
        /// Наскрізний номер канону (Буття — 10 … Об'явлення — 730); 0 — невідомо.
        final int canon;
        final String name;
        final String shortName;
        final int chapters;

        Book(int index, int canon, String name, String shortName, int chapters) {
            this.index = index;
            this.canon = canon;
            this.name = name;
            this.shortName = shortName;
            this.chapters = chapters;
        }
    }

    static final class Verse {
        final int number;
        final String text;

        Verse(int number, String text) {
            this.number = number;
            this.text = text;
        }
    }

    static final class Song {
        final int index;
        final String title;

        Song(int index, String title) {
            this.index = index;
            this.title = title;
        }
    }

    static final class Part {
        final int index;
        final String kind;
        final String text;

        Part(int index, String kind, String text) {
            this.index = index;
            this.kind = kind;
            this.text = text;
        }
    }

    Library(Context context) {
        super(context.getApplicationContext(), "library.db", null, 1);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL("CREATE TABLE modules (id TEXT PRIMARY KEY, kind TEXT NOT NULL, name TEXT NOT NULL,"
            + " short TEXT NOT NULL DEFAULT '', source TEXT NOT NULL DEFAULT '',"
            + " original TEXT NOT NULL DEFAULT '', added INTEGER NOT NULL DEFAULT 0)");
        db.execSQL("CREATE TABLE books (module TEXT NOT NULL, idx INTEGER NOT NULL, canon INTEGER NOT NULL DEFAULT 0,"
            + " name TEXT NOT NULL, short TEXT NOT NULL DEFAULT '', chapters INTEGER NOT NULL DEFAULT 0,"
            + " PRIMARY KEY (module, idx))");
        db.execSQL("CREATE TABLE verses (module TEXT NOT NULL, book INTEGER NOT NULL, chapter INTEGER NOT NULL,"
            + " verse INTEGER NOT NULL, text TEXT NOT NULL)");
        db.execSQL("CREATE INDEX verses_at ON verses (module, book, chapter)");
        db.execSQL("CREATE TABLE songs (module TEXT NOT NULL, idx INTEGER NOT NULL, title TEXT NOT NULL,"
            + " fold TEXT NOT NULL, PRIMARY KEY (module, idx))");
        db.execSQL("CREATE TABLE parts (module TEXT NOT NULL, song INTEGER NOT NULL, idx INTEGER NOT NULL,"
            + " kind TEXT NOT NULL, text TEXT NOT NULL)");
        db.execSQL("CREATE INDEX parts_at ON parts (module, song)");
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int from, int to) { }

    // MARK: Читання

    List<Module> modules(String kind) {
        List<Module> result = new ArrayList<>();
        try (Cursor c = getReadableDatabase().rawQuery(
                "SELECT id, kind, name, short, source, original FROM modules WHERE kind = ? ORDER BY name COLLATE NOCASE",
                new String[] { kind })) {
            while (c.moveToNext()) {
                result.add(new Module(c.getString(0), c.getString(1), c.getString(2), c.getString(3),
                    c.getString(4), c.getString(5)));
            }
        }
        return result;
    }

    Module module(String id) {
        try (Cursor c = getReadableDatabase().rawQuery(
                "SELECT id, kind, name, short, source, original FROM modules WHERE id = ?", new String[] { id })) {
            return c.moveToNext()
                ? new Module(c.getString(0), c.getString(1), c.getString(2), c.getString(3), c.getString(4), c.getString(5))
                : null;
        }
    }

    List<Book> books(String module) {
        List<Book> result = new ArrayList<>();
        try (Cursor c = getReadableDatabase().rawQuery(
                "SELECT idx, canon, name, short, chapters FROM books WHERE module = ? ORDER BY idx",
                new String[] { module })) {
            while (c.moveToNext()) {
                result.add(new Book(c.getInt(0), c.getInt(1), c.getString(2), c.getString(3), c.getInt(4)));
            }
        }
        return result;
    }

    /// Номери розділів книги — ті, що справді є в базі: у модулях буває й
    /// нульовий розділ, і менше розділів, ніж обіцяє `ChapterQty`.
    List<Integer> chapters(String module, int book) {
        List<Integer> result = new ArrayList<>();
        try (Cursor c = getReadableDatabase().rawQuery(
                "SELECT DISTINCT chapter FROM verses WHERE module = ? AND book = ? ORDER BY chapter",
                new String[] { module, String.valueOf(book) })) {
            while (c.moveToNext()) result.add(c.getInt(0));
        }
        return result;
    }

    List<Verse> verses(String module, int book, int chapter) {
        List<Verse> result = new ArrayList<>();
        try (Cursor c = getReadableDatabase().rawQuery(
                "SELECT verse, text FROM verses WHERE module = ? AND book = ? AND chapter = ? ORDER BY verse",
                new String[] { module, String.valueOf(book), String.valueOf(chapter) })) {
            while (c.moveToNext()) result.add(new Verse(c.getInt(0), c.getString(1)));
        }
        return result;
    }

    /// Пісні за номером або словами назви. Порожній фільтр — усі по порядку.
    List<Song> songs(String module, String filter, int limit) {
        List<Song> result = new ArrayList<>();
        String wanted = fold(filter);
        String sql;
        String[] args;
        if (wanted.isEmpty()) {
            sql = "SELECT idx, title FROM songs WHERE module = ? ORDER BY idx LIMIT " + limit;
            args = new String[] { module };
        } else if (wanted.matches("\\d+")) {
            sql = "SELECT idx, title FROM songs WHERE module = ? AND (idx = ? OR fold LIKE ?) ORDER BY idx LIMIT " + limit;
            args = new String[] { module, String.valueOf(Integer.parseInt(wanted) - 1), "%" + wanted + "%" };
        } else {
            sql = "SELECT idx, title FROM songs WHERE module = ? AND fold LIKE ? ORDER BY idx LIMIT " + limit;
            args = new String[] { module, "%" + wanted + "%" };
        }
        try (Cursor c = getReadableDatabase().rawQuery(sql, args)) {
            while (c.moveToNext()) result.add(new Song(c.getInt(0), c.getString(1)));
        }
        return result;
    }

    List<Part> parts(String module, int song) {
        List<Part> result = new ArrayList<>();
        try (Cursor c = getReadableDatabase().rawQuery(
                "SELECT idx, kind, text FROM parts WHERE module = ? AND song = ? ORDER BY idx",
                new String[] { module, String.valueOf(song) })) {
            while (c.moveToNext()) result.add(new Part(c.getInt(0), c.getString(1), c.getString(2)));
        }
        return result;
    }

    /// Прибрати модуль — і рядки, і збережений файл.
    void delete(String id) {
        Module known = module(id);
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            deleteRows(db, id);
            db.delete("modules", "id = ?", new String[] { id });
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
        }
        if (known != null && !known.original.isEmpty()) {
            //noinspection ResultOfMethodCallIgnored
            new File(known.original).delete();
        }
    }

    private static void deleteRows(SQLiteDatabase db, String id) {
        String[] args = { id };
        db.delete("books", "module = ?", args);
        db.delete("verses", "module = ?", args);
        db.delete("songs", "module = ?", args);
        db.delete("parts", "module = ?", args);
    }

    /// Рядок для пошуку: малі літери, «ё» як «е», без розділових знаків.
    static String fold(String text) {
        if (text == null) return "";
        String lower = text.toLowerCase(Locale.ROOT).replace('ё', 'е');
        StringBuilder out = new StringBuilder(lower.length());
        boolean space = false;
        for (int i = 0; i < lower.length(); i++) {
            char ch = lower.charAt(i);
            if (Character.isLetterOrDigit(ch)) {
                out.append(ch);
                space = false;
            } else if (!space && out.length() > 0) {
                out.append(' ');
                space = true;
            }
        }
        return out.toString().trim();
    }

    // MARK: Запис

    /// Запис модуля — одна транзакція: або модуль ліг увесь, або його немає.
    /// Попередня версія того самого модуля при цьому замінюється.
    Writer write(String id, String kind, String name, String shortName, String source, String original) {
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        deleteRows(db, id);
        ContentValues row = new ContentValues();
        row.put("id", id);
        row.put("kind", kind);
        row.put("name", name == null || name.isEmpty() ? id : name);
        row.put("short", shortName == null ? "" : shortName);
        row.put("source", source);
        row.put("original", original == null ? "" : original);
        row.put("added", System.currentTimeMillis());
        db.insertWithOnConflict("modules", null, row, SQLiteDatabase.CONFLICT_REPLACE);
        return new Writer(db, id);
    }

    static final class Writer {
        private final SQLiteDatabase db;
        private final String id;
        private final SQLiteStatement book;
        private final SQLiteStatement verse;
        private final SQLiteStatement song;
        private final SQLiteStatement part;
        private boolean finished;
        int verses;
        int songs;

        private Writer(SQLiteDatabase db, String id) {
            this.db = db;
            this.id = id;
            book = db.compileStatement("INSERT OR REPLACE INTO books (module, idx, canon, name, short, chapters) VALUES (?, ?, ?, ?, ?, ?)");
            verse = db.compileStatement("INSERT INTO verses (module, book, chapter, verse, text) VALUES (?, ?, ?, ?, ?)");
            song = db.compileStatement("INSERT OR REPLACE INTO songs (module, idx, title, fold) VALUES (?, ?, ?, ?)");
            part = db.compileStatement("INSERT INTO parts (module, song, idx, kind, text) VALUES (?, ?, ?, ?, ?)");
        }

        void book(int index, int canon, String name, String shortName, int chapters) {
            book.bindString(1, id);
            book.bindLong(2, index);
            book.bindLong(3, canon);
            book.bindString(4, name == null ? "" : name);
            book.bindString(5, shortName == null ? "" : shortName);
            book.bindLong(6, chapters);
            book.executeInsert();
        }

        void verse(int bookIndex, int chapter, int number, String text) {
            verse.bindString(1, id);
            verse.bindLong(2, bookIndex);
            verse.bindLong(3, chapter);
            verse.bindLong(4, number);
            verse.bindString(5, text);
            verse.executeInsert();
            verses++;
        }

        void song(int index, String title) {
            song.bindString(1, id);
            song.bindLong(2, index);
            song.bindString(3, title);
            song.bindString(4, (index + 1) + " " + fold(title));
            song.executeInsert();
            songs++;
        }

        void part(int songIndex, int index, String kind, String text) {
            part.bindString(1, id);
            part.bindLong(2, songIndex);
            part.bindLong(3, index);
            part.bindString(4, kind == null ? "" : kind);
            part.bindString(5, text == null ? "" : text);
            part.executeInsert();
        }

        void commit() {
            if (finished) return;
            finished = true;
            db.setTransactionSuccessful();
            db.endTransaction();
        }

        void abort() {
            if (finished) return;
            finished = true;
            db.endTransaction();
        }
    }
}

File: Проповідник/Sources/Library.cs
// =============================================================================
//  Library.cs — своя бібліотека перекладів і пісенників
// =============================================================================
//  Перенесено з планшета (Library.java). Звідки б модуль не прийшов — зі
//  «Слова» по мережі, з GitHub чи з файла, — він лягає в одну базу однаковими
//  рядками: книга, розділ, вірш, текст. Екран плану тому не знає, якого
//  формату був модуль.
//
//  Разом із модулем з GitHub чи з файла зберігається й сам файл (Original):
//  якщо в програмі на служінні такого перекладу не виявиться, «Проповідник»
//  відвезе їй саме його, і програма поставить модуль собі.
// =============================================================================

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using Microsoft.Data.Sqlite;

namespace Propovidnyk;

public sealed record Module(string Id, string Kind, string Name, string ShortName, string Source, string Original);
public sealed record Book(int Index, int Canon, string Name, string ShortName, int Chapters);
public sealed record Verse(int Number, string Text);
public sealed record Song(int Index, string Title);
public sealed record Part(int Index, string Kind, string Text);

public sealed class Library : IDisposable
{
    public const string Bible = "bible";
    public const string Songs = "songs";

    readonly SqliteConnection _db;

    public Library(string? path = null)
    {
        _db = new SqliteConnection(new SqliteConnectionStringBuilder
        {
            DataSource = path ?? Paths.Library,
            Mode = SqliteOpenMode.ReadWriteCreate,
        }.ToString());
        _db.Open();
        Exec("PRAGMA journal_mode=WAL");
        Exec("""
            CREATE TABLE IF NOT EXISTS modules (id TEXT PRIMARY KEY, kind TEXT NOT NULL, name TEXT NOT NULL,
              short TEXT NOT NULL DEFAULT '', source TEXT NOT NULL DEFAULT '',
              original TEXT NOT NULL DEFAULT '', added INTEGER NOT NULL DEFAULT 0);
            CREATE TABLE IF NOT EXISTS books (module TEXT NOT NULL, idx INTEGER NOT NULL, canon INTEGER NOT NULL DEFAULT 0,
              name TEXT NOT NULL, short TEXT NOT NULL DEFAULT '', chapters INTEGER NOT NULL DEFAULT 0,
              PRIMARY KEY (module, idx));
            CREATE TABLE IF NOT EXISTS verses (module TEXT NOT NULL, book INTEGER NOT NULL, chapter INTEGER NOT NULL,
              verse INTEGER NOT NULL, text TEXT NOT NULL);
            CREATE INDEX IF NOT EXISTS verses_at ON verses (module, book, chapter);
            CREATE TABLE IF NOT EXISTS songs (module TEXT NOT NULL, idx INTEGER NOT NULL, title TEXT NOT NULL,
              fold TEXT NOT NULL, PRIMARY KEY (module, idx));
            CREATE TABLE IF NOT EXISTS parts (module TEXT NOT NULL, song INTEGER NOT NULL, idx INTEGER NOT NULL,
              kind TEXT NOT NULL, text TEXT NOT NULL);
            CREATE INDEX IF NOT EXISTS parts_at ON parts (module, song);
            """);
    }

    public void Dispose() => _db.Dispose();

    void Exec(string sql)
    {
        using var command = _db.CreateCommand();
        command.CommandText = sql;
        command.ExecuteNonQuery();
    }

    SqliteCommand Query(string sql, params object[] args)
    {
        var command = _db.CreateCommand();
        command.CommandText = sql;
        for (var i = 0; i < args.Length; i++) command.Parameters.AddWithValue("$" + (i + 1), args[i]);
        return command;
    }

    // MARK: Читання

    public List<Module> Modules(string kind)
    {
        var result = new List<Module>();
        using var command = Query("SELECT id, kind, name, short, source, original FROM modules WHERE kind = $1 ORDER BY name COLLATE NOCASE", kind);
        using var r = command.ExecuteReader();
        while (r.Read()) result.Add(new Module(r.GetString(0), r.GetString(1), r.GetString(2), r.GetString(3), r.GetString(4), r.GetString(5)));
        return result;
    }

    public Module? Find(string id)
    {
        using var command = Query("SELECT id, kind, name, short, source, original FROM modules WHERE id = $1", id);
        using var r = command.ExecuteReader();
        return r.Read() ? new Module(r.GetString(0), r.GetString(1), r.GetString(2), r.GetString(3), r.GetString(4), r.GetString(5)) : null;
    }

    public List<Book> Books(string module)
    {
        var result = new List<Book>();
        using var command = Query("SELECT idx, canon, name, short, chapters FROM books WHERE module = $1 ORDER BY idx", module);
        using var r = command.ExecuteReader();
        while (r.Read()) result.Add(new Book(r.GetInt32(0), r.GetInt32(1), r.GetString(2), r.GetString(3), r.GetInt32(4)));
        return result;
    }

    /// Номери розділів — ті, що справді є в базі: буває й нульовий розділ.
    public List<int> Chapters(string module, int book)
    {
        var result = new List<int>();
        using var command = Query("SELECT DISTINCT chapter FROM verses WHERE module = $1 AND book = $2 ORDER BY chapter", module, book);
        using var r = command.ExecuteReader();
        while (r.Read()) result.Add(r.GetInt32(0));
        return result;
    }

    public List<Verse> Verses(string module, int book, int chapter)
    {
        var result = new List<Verse>();
        using var command = Query("SELECT verse, text FROM verses WHERE module = $1 AND book = $2 AND chapter = $3 ORDER BY verse", module, book, chapter);
        using var r = command.ExecuteReader();
        while (r.Read()) result.Add(new Verse(r.GetInt32(0), r.GetString(1)));
        return result;
    }

    /// Пісні за номером або словами назви. Порожній фільтр — усі по порядку.
    public List<Song> SongsOf(string module, string filter, int limit = 500)
    {
        var result = new List<Song>();
        var wanted = Fold(filter);
        SqliteCommand command;
        if (wanted.Length == 0)
            command = Query($"SELECT idx, title FROM songs WHERE module = $1 ORDER BY idx LIMIT {limit}", module);
        else if (wanted.All(char.IsDigit) && int.TryParse(wanted, out var number))
            command = Query($"SELECT idx, title FROM songs WHERE module = $1 AND (idx = $2 OR fold LIKE $3) ORDER BY idx LIMIT {limit}",
                            module, number - 1, "%" + wanted + "%");
        else
            command = Query($"SELECT idx, title FROM songs WHERE module = $1 AND fold LIKE $2 ORDER BY idx LIMIT {limit}", module, "%" + wanted + "%");
        using (command)
        using (var r = command.ExecuteReader())
            while (r.Read()) result.Add(new Song(r.GetInt32(0), r.GetString(1)));
        return result;
    }

    public List<Part> Parts(string module, int song)
    {
        var result = new List<Part>();
        using var command = Query("SELECT idx, kind, text FROM parts WHERE module = $1 AND song = $2 ORDER BY idx", module, song);
        using var r = command.ExecuteReader();
        while (r.Read()) result.Add(new Part(r.GetInt32(0), r.GetString(1), r.GetString(2)));
        return result;
    }

    public int Count(string table, string module)
    {
        using var command = Query($"SELECT COUNT(*) FROM {table} WHERE module = $1", module);
        return Convert.ToInt32(command.ExecuteScalar());
    }

    /// Прибрати модуль — і рядки, і збережений файл.
    public void Delete(string id)
    {
        var known = Find(id);
        using (var transaction = _db.BeginTransaction())
        {
            DeleteRows(id, transaction);
            using var command = Query("DELETE FROM modules WHERE id = $1", id);
            command.Transaction = transaction;
            command.ExecuteNonQuery();
            transaction.Commit();
        }
        if (known != null && known.Original.Length > 0)
        {
            try { File.Delete(known.Original); } catch { /* файл уже прибрали */ }
        }
    }

    void DeleteRows(string id, SqliteTransaction transaction)
    {
        foreach (var table in new[] { "books", "verses", "songs", "parts" })
        {
            using var command = Query($"DELETE FROM {table} WHERE module = $1", id);
            command.Transaction = transaction;
            command.ExecuteNonQuery();
        }
    }

    /// Рядок для пошуку: малі літери, «ё» як «е», без розділових знаків.
    public static string Fold(string? text)
    {
        if (string.IsNullOrEmpty(text)) return "";
        var lower = text.ToLowerInvariant().Replace('ё', 'е');
        var output = new StringBuilder(lower.Length);
        var space = false;
        foreach (var ch in lower)
        {
            if (char.IsLetterOrDigit(ch))
            {
                output.Append(ch);
                space = false;
            }
            else if (!space && output.Length > 0)
            {
                output.Append(' ');
                space = true;
            }
        }
        return output.ToString().Trim();
    }

    // MARK: Запис

    /// Запис модуля — одна транзакція: або модуль ліг увесь, або його немає.
    public Writer Write(string id, string kind, string name, string shortName, string source, string original)
    {
        var transaction = _db.BeginTransaction();
        DeleteRows(id, transaction);
        using (var command = Query("INSERT OR REPLACE INTO modules (id, kind, name, short, source, original, added) VALUES ($1, $2, $3, $4, $5, $6, $7)",
                                   id, kind, string.IsNullOrEmpty(name) ? id : name, shortName ?? "", source, original ?? "",
                                   DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()))
        {
            command.Transaction = transaction;
            command.ExecuteNonQuery();
        }
        return new Writer(this, id, transaction);
    }

    public sealed class Writer
    {
        readonly Library _library;
        readonly string _id;
        readonly SqliteTransaction _transaction;
        readonly SqliteCommand _book, _verse, _song, _part;
        bool _finished;
        public int VerseCount { get; private set; }
        public int SongCount { get; private set; }

        internal Writer(Library library, string id, SqliteTransaction transaction)
        {
            _library = library;
            _id = id;
            _transaction = transaction;
            _book = Prepare("INSERT OR REPLACE INTO books (module, idx, canon, name, short, chapters) VALUES ($1, $2, $3, $4, $5, $6)", 6);
            _verse = Prepare("INSERT INTO verses (module, book, chapter, verse, text) VALUES ($1, $2, $3, $4, $5)", 5);
            _song = Prepare("INSERT OR REPLACE INTO songs (module, idx, title, fold) VALUES ($1, $2, $3, $4)", 4);
            _part = Prepare("INSERT INTO parts (module, song, idx, kind, text) VALUES ($1, $2, $3, $4, $5)", 5);
        }

        SqliteCommand Prepare(string sql, int count)
        {
            var command = _library._db.CreateCommand();
            command.CommandText = sql;
            command.Transaction = _transaction;
            for (var i = 1; i <= count; i++) command.Parameters.Add(new SqliteParameter("$" + i, null));
            command.Prepare();
            return command;
        }

        static void Run(SqliteCommand command, params object[] values)
        {
            for (var i = 0; i < values.Length; i++) command.Parameters[i].Value = values[i];
            command.ExecuteNonQuery();
        }

        public void Book(int index, int canon, string? name, string? shortName, int chapters) =>
            Run(_book, _id, index, canon, name ?? "", shortName ?? "", chapters);

        public void Verse(int book, int chapter, int number, string text)
        {
            Run(_verse, _id, book, chapter, number, text);
            VerseCount++;
        }

        public void Song(int index, string title)
        {
            Run(_song, _id, index, title, (index + 1) + " " + Fold(title));
            SongCount++;
        }

        public void Part(int song, int index, string? kind, string? text) =>
            Run(_part, _id, song, index, kind ?? "", text ?? "");

        public void Commit()
        {
            if (_finished) return;
            _finished = true;
            _transaction.Commit();
            Dispose();
        }

        public void Abort()
        {
            if (_finished) return;
            _finished = true;
            _transaction.Rollback();
            Dispose();
        }

        void Dispose()
        {
            _book.Dispose(); _verse.Dispose(); _song.Dispose(); _part.Dispose();
            _transaction.Dispose();
        }
    }
}
